### PR TITLE
feat: Added support for Authorizers

### PR DESCRIPTION
**Pull request by bobbydeveaux**
## Description
Adding a new section `Authorizers` which allows anyone to create multiple authorisers of varying types.

## Motivation and Context
We needed to add a JWT and OIDC Auth to our project API Gateway

## How Has This Been Tested?
Deployed into our own sandbox account and working great.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -58,6 +58,72 @@ module "api_gateway" {
...
+  default_stage_access_log_destination_arn = "arn:aws:logs:eu-west-1:835367859851:log-group:debug-apigateway"
+  default_stage_access_log_format          = "$context.identity.sourceIp - - [$context.requestTime] \"$context.httpMethod $context.routeKey $context.protocol\" $context.status $context.responseLength $context.requestId $context.integrationErrorMessage"
+
+  authorizers ={
+
+    "azure" = {
+      authorizer_type  = "JWT"
+      identity_sources = "$request.header.Authorization"
+      name             = "azure-auth"
+      audience         = var.azure_client_id
```
Comment (antonbabenko) [inline README.md:91]: Replace `var.azure_client_id` and other `var.` references with some string value that looks similar to the real value (e.g. `skjdgbsadgkjfbnsdexample`)

File: main.tf [hunk 2]
```diff
@@ -123,7 +123,7 @@ resource "aws_apigatewayv2_route" "this" {
 
   api_key_required                    = lookup(each.value, "api_key_required", null)
   authorization_type                  = lookup(each.value, "authorization_type", "NONE")
-  authorizer_id                       = lookup(each.value, "authorizer_id", null)
+  authorizer_id                       = try(aws_apigatewayv2_authorizer.this[lookup(each.value, "authorizer_key", null)].id, lookup(each.value, "authorizer_id", null))
```
Comment (antonbabenko) [inline main.tf:126]:
```suggestion
  authorizer_id                       = try(aws_apigatewayv2_authorizer.this[each.value.authorizer_key].id, each.value.authorizer_id, null)
```

File: main.tf [hunk 3]
```diff
@@ -175,6 +177,27 @@ resource "aws_apigatewayv2_integration" "this" {
...
 }
 
+# Authorizers
+resource "aws_apigatewayv2_authorizer" "this" {
+  for_each = var.create && var.create_routes_and_integrations ? var.authorizers : {}
+
+  api_id    = aws_apigatewayv2_api.this[0].id
+
+  authorizer_type  = lookup(each.value, "authorizer_type", null)
+  identity_sources = [lookup(each.value, "identity_sources", null)]
```
Comment (antonbabenko) [inline main.tf:187]:
```suggestion
  identity_sources = try(flatten(tolist(each.value.identity_sources)), null)
```

File: main.tf [hunk 4]
```diff
@@ -175,6 +177,27 @@ resource "aws_apigatewayv2_integration" "this" {
...
+  name             = lookup(each.value, "name", null)
+  authorizer_uri   = lookup(each.value, "authorizer_uri", null)
+  authorizer_payload_format_version = lookup(each.value, "authorizer_payload_format_version", null)
+
+  dynamic "jwt_configuration" {
+    for_each = lookup(each.value, "audience", null) == null ? tolist([]) : tolist([lookup(each.value, "audience", null)])
```
Comment (antonbabenko) [inline main.tf:193]:
```suggestion
    for_each = try(flatten(tolist(each.value.audience)), [])
```
or similar

File: main.tf [hunk 5]
```diff
@@ -175,6 +177,27 @@ resource "aws_apigatewayv2_integration" "this" {
...
+    content {
+      audience = [lookup(each.value, "audience", null)]
```
Comment (antonbabenko) [inline main.tf:195]:
```suggestion
      audience = jwt_configuration.value
```

**Review by antonbabenko — CHANGES_REQUESTED** [review]
Looks pretty good.

Please update an example in `examples/complete-http` to use the feature you are adding in this PR.

File: main.tf (hunk 5 above)
Comment (bobbydeveaux) [reply main.tf:195]: TBH - Both sub-elements (issuer and audience are both optional - the for_each should be based on the either of them being present, but not necessarily both...

Comment (bobbydeveaux) [reply main.tf:195]: I made this change assuming that audience and issuer would both be present. But it's possible to supply just issuer.

Comment (antonbabenko) [reply main.tf:195]: Then let's change `for_each` on line 193 to handle that at least one value (audience or issuer) should be present.

**Comment by antonbabenko** [comment]
Thank you @bobbydeveaux for the contribution!

I have updated the code in examples to make it executable.

**Comment by antonbabenko** [comment]
This PR is included in version 1.6.0 :tada:

**Pull request merged**